### PR TITLE
New encryption service

### PR DESCRIPTION
**Pull request by gonzaloriestra**
Related to: https://github.com/CartoDB/product/issues/291

New gem added to use Argon2 encryption: https://github.com/technion/ruby-argon2

The service is not used yet, but I've checked the compatibility with current encryptions in CartoDB in the tests. It will be used everywhere in the next PR.

### CODE REVIEW

File: app/services/carto/encryption_service.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  class EncryptionService
+
+    AUTH_DIGEST_KEY_SHA1 = '47f940ec20a0993b5e9e4310461cc8a6a7fb84e3'.freeze
+    AUTH_DIGEST_KEY_SHA256 = '1211b3e77138f6e1724721f1ab740c9c70e66ba6fec5e989bb6640c4541ed15d06dbd5fdcbd3052b'.freeze
+    AUTH_TOKEN_DIGEST_KEY = '6da98b2da1b38c5ada2547ad2c3268caa1eb58dc20c9144ead844a2eda1917067a06dcb54833ba2'.freeze
+
+    DEFAULT_SHA_CLASS = Digest::SHA1
+
+    def encrypt(password:, sha_class: nil, salt: nil)
+      return Argon2::Password.create(password) unless sha_class
```
Comment (gonzaloriestra) [inline app/services/carto/encryption_service.rb]: This is using the default parameters: `t_cost: 2, m_cost = 16`. It could be increased up to 750 and 31, respectively.

According to the docs, `m_cost` could be increased for more security. But I've noticed that it's much slower and it throws memory errors locally when increasing it too much. So maybe it's better to keep it with the default values. Thoughs?

Comment (javitonino) [reply app/services/carto/encryption_service.rb]: The default should be pretty decent I think. Increasing the cost exponentially increases memory and CPU cost (`m_cost` is memory cost, `time_cost` is time cost).

> Here the time cost t_cost is set to **2** iterations, the memory cost m_cost is set to 2^**16** kibibytes (64 mebibytes), and parallelism is set to 1 (single-thread).

Looks like reasonable values to me. Let's trust the cryptographers.

File: Gemfile.lock [hunk 2]
```diff
@@ -405,13 +414,13 @@ GEM
...
-    warden (1.2.6)
+    warden (1.2.7)
       rack (>= 1.0)
     webrick (1.3.1)
-    websocket (1.2.2)
+    websocket (1.2.8)
     xpath (0.1.4)
       nokogiri (~> 1.3)
-    zeus (0.15.13)
+    zeus (0.15.14)
```
Comment (javitonino) [inline Gemfile.lock]: Some crazy updates in the lockfile. Are you confident that this won't blow up in our faces?

Not opposed to the updates but I'd do them in a separate PR.

File: app/services/carto/encryption_service.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
+module Carto
+  class EncryptionService
+
+    AUTH_DIGEST_KEY_SHA1 = '47f940ec20a0993b5e9e4310461cc8a6a7fb84e3'.freeze
+    AUTH_DIGEST_KEY_SHA256 = '1211b3e77138f6e1724721f1ab740c9c70e66ba6fec5e989bb6640c4541ed15d06dbd5fdcbd3052b'.freeze
+    AUTH_TOKEN_DIGEST_KEY = '6da98b2da1b38c5ada2547ad2c3268caa1eb58dc20c9144ead844a2eda1917067a06dcb54833ba2'.freeze
```
Comment (javitonino) [inline app/services/carto/encryption_service.rb]: Similar to this, `Argon2` has a key parameter which is some fixed string added to the hashing algorithm (also called pepper). I'd add support for it and make it configurable from `app_config.yml`.

File: app/services/carto/encryption_service.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    end
+
+    def decode_hex(hex_string)
+      hex_string.scan(/../).map(&:hex).pack('c*')
+    end
+
+    def verify_sha(password, secure_password, salt)
+      case secure_password
+      when /^\h{40}$/ then secure_password == sha_digest(sha_class: Digest::SHA1, args: [salt, password])
+      when /^\h{64}$/ then secure_password == sha_digest(sha_class: Digest::SHA256, args: [salt, password])
```
Comment (javitonino) [inline app/services/carto/encryption_service.rb]: If we do a migration from (password + salt) fields to a single password field (with $algorithm$salt$crypted), we could make this case much simpler (just split by $ and check the first one).

Maybe worth it to do that first if we are going to do it.

File: app/services/carto/encryption_service.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      args_join = '--' if sha_class == Digest::SHA1
+
+      10.times do
+        joined_args = [digest, args, digest_key].flatten.join(args_join)
+        digest = sha_class.hexdigest(joined_args)
+      end
+      digest
+    end
+
+    def default_digest_key(sha_class)
```
Comment (javitonino) [inline app/services/carto/encryption_service.rb]: You could do a hash using the class object as a key:
```
DIGEST_TOKENS = { Digest::SHA1 => 'aaa', Digest::SHA256 => 'bbb' }.freeze
token = DIGEST_TOKENS[sha_class]
```

File: spec/services/carto/encryption_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    end
+
+    it "returns a different output each time" do
+      result1 = @service.encrypt(password: "wadus")
+      result2 = @service.encrypt(password: "wadus")
+      result1.should_not eql result2
+    end
+
+    it "allows to use SHA1" do
+      result = @service.encrypt(password: "wadus", sha_class: Digest::SHA1, salt: "himalayan")
```
Comment (javitonino) [inline spec/services/carto/encryption_service_spec.rb]: LOL

File: app/services/carto/encryption_service.rb [hunk 7]
```diff
@@ -8,28 +8,52 @@ class EncryptionService
...
 
+    def argon2?(encryption)
+      encryption =~ /^\h{194}$/
+    end
+
     private
 
+    def argon2_encrypt(password)
+      argon_hash = Argon2::Password.create(password)
+      encode_to_hex(argon_hash)
```
Comment (javitonino) [inline app/services/carto/encryption_service.rb]: Why encode to hex?

File: Gemfile.lock (hunk 2 above)
Comment (gonzaloriestra) [reply Gemfile.lock]: Oops, this was not intended. Now I remember I had some problems locally and I ended up removing the file, so it updated everything. I'll revert it to add only the required stuff.

File: app/services/carto/encryption_service.rb (hunk 7 above)
Comment (gonzaloriestra) [reply app/services/carto/encryption_service.rb]: The password for the user database is generated this way:
```
def database_password
  @user.crypted_password + database_username
end
```
Until now, the encrypted password was only hex characters, but Argon2 includes other chars like `+`, `/` or `$`, and it breaks it.

Also, we may want to use to generate random tokens with Argon2 to be used API calls or emails, and it would break as well.

File: app/services/carto/encryption_service.rb (hunk 3 above)
Comment (gonzaloriestra) [reply app/services/carto/encryption_service.rb]: I didn't use it because I think it doesn't make much sense if it is hardcoded in the public repo, as we do now... But yeah, we can add it to the private configuration :+1:

File: app/services/carto/encryption_service.rb (hunk 7 above)
Comment (javitonino) [reply app/services/carto/encryption_service.rb]: Ains, I hate when these things happen xD

If we must, I prefer doing the hack here, in this method. Something like, if the `crypted_password` contains unsupported characters, encode it to hex for the purposes of database_password (maybe just take a hash of it, since argon2 passwords can get quite long).

The proper solution would be to just store the db password in the database and generate it completely at random.

> Also, we may want to use to generate random tokens with Argon2 to be used API calls or emails, and it would break as well.

I'd say this is a feature. This is something we definitely do not want to do. Anything derived from a password hashing algorithm should not be exposed to the public.

Comment (gonzaloriestra) [reply app/services/carto/encryption_service.rb]: You are right, let's implement your first suggestion then.

File: app/services/carto/encryption_service.rb (hunk 4 above)
Comment (gonzaloriestra) [reply app/services/carto/encryption_service.rb]: I think it will be easier if we do it later when new passwords are already using Argon2. Otherwise, we would have to change twice the way new passwords are saved (first to save them as `$sha1$salt$crypted` and then with Argon2) and enable verification with old and new format.

Is it ok? We'll do it, I promise :slightly_smiling_face:

Comment (javitonino) [reply app/services/carto/encryption_service.rb]: Sure. However you think it's best! You are the one which has the more up to date context on how all of this holds together :D

File: Gemfile [hunk 8]
```diff
@@ -67,6 +67,7 @@ gem 'gibbon',                  '1.1.4'
 # Service components (/services)
 gem 'virtus',                   '1.0.5'
 gem 'uuidtools',                '2.1.5'
+gem 'argon2',                   '~> 2.0.2'
```
Comment (javitonino) [inline Gemfile]: I'd pin this to `~> 2.0` (which is `>=2 <3`). Looking at the gem maintainer, they are quite good with following SemVer and we probably want minors to be automatically updated (e.g: algorithm default changes).

File: app/services/carto/encryption_service.rb [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      Digest::SHA1 => '47f940ec20a0993b5e9e4310461cc8a6a7fb84e3',
+      Digest::SHA256 => '1211b3e77138f6e1724721f1ab740c9c70e66ba6fec5e989bb6640c4541ed15d06dbd5fdcbd3052b'
+    }.freeze
+
+    DEFAULT_SHA_CLASS = Digest::SHA1
+
+    def encrypt(password:, sha_class: nil, salt: nil, secret: nil)
+      return encrypt_argon2(password, secret) unless sha_class
+      encrypt_sha(sha_class: sha_class, args: [salt, password])
+    end
```
Comment (javitonino) [inline app/services/carto/encryption_service.rb]: This method might be easier to read with a straight forward `if sha_class`, but I don't think it's worth changing if we are not doing more changes.

Anyway, we might remove this later (leave only verification code for old algorithms).

Comment (gonzaloriestra) [reply app/services/carto/encryption_service.rb]: We won't be able to remove it yet because we need it to create tokens from emails, but maybe we can change that system also.
